# Unit 12 - Tales from the Crypto

---


## 1. Sentiment Analysis

Use the [newsapi](https://newsapi.org/) to pull the latest news articles for Bitcoin and Ethereum and create a DataFrame of sentiment scores for each coin.

Use descriptive statistics to answer the following questions:
1. Which coin had the highest mean positive score?
2. Which coin had the highest negative score?
3. Which coin had the highest positive score?

In [ ]:
# Initial imports
import os
import pandas as pd
from newsapi import NewsApiClient
from dotenv import load_dotenv
import nltk as nltk
nltk.download('vader_lexicon')
from nltk.sentiment.vader import SentimentIntensityAnalyzer
analyzer = SentimentIntensityAnalyzer()

%matplotlib inline

[nltk_data] Downloading package vader_lexicon to
[nltk_data]     C:\Users\Ldkel\AppData\Roaming\nltk_data...
[nltk_data]   Package vader_lexicon is already up-to-date!


In [ ]:
# Read your api key environment variable

news_api_key = "5311671e6a8941878360f7a695b22582"

In [58]:
# Create a newsapi client
newsapi = NewsApiClient(api_key=news_api_key)

In [59]:
# Fetch the Bitcoin news articles
btc_articles = newsapi.get_everything(q="bitcoin", language="en", sort_by="relevancy")
btc_articles

{'status': 'ok',
 'totalResults': 9402,
 'articles': [{'source': {'id': 'engadget', 'name': 'Engadget'},
   'author': 'Kris Holt',
   'title': 'New York passes a bill to limit bitcoin mining',
   'description': "New York lawmakers have passed a bill\r\n that would temporarily ban new bitcoin\r\n mining operations. Early on Friday, state senators voted 36-27 to pass the legislation. It's now bound for the desk of Governor Kathy Hochul, who will sign it into law or veto th…",
   'url': 'https://www.engadget.com/new-york-cryptocurrency-bill-bitcoin-mining-climate-change-161126292.html',
   'urlToImage': 'https://s.yimg.com/os/creatr-uploaded-images/2021-05/a8217250-bdfa-11eb-bfc4-2663225cea83',
   'publishedAt': '2022-06-03T16:11:26Z',
   'content': "New York lawmakers have passed a bill\r\n that would temporarily ban new bitcoin\r\n mining operations. Early on Friday, state senators voted 36-27 to pass the legislation. It's now bound for the desk of… [+2036 chars]"},
  {'source': {'id': 

In [60]:
# Fetch the Ethereum news articles
eth_articles = newsapi.get_everything(q="ethereum", language="en", sort_by="relevancy")
eth_articles

{'status': 'ok',
 'totalResults': 4556,
 'articles': [{'source': {'id': None, 'name': 'Gizmodo.com'},
   'author': 'Kyle Barr',
   'title': 'GameStop Dunks Its Head Into the Crypto Kiddie Pool',
   'description': 'GameStop has officially thrown itself headlong into the web3 vipers nest with a new app release, though it’s hard to say whether its proposed population of gamers and game developers will take up the company on its belated, head-first jump into the crypto sph…',
   'url': 'https://gizmodo.com/gamestop-crypto-nft-wallet-blockchain-1848965386',
   'urlToImage': 'https://i.kinja-img.com/gawker-media/image/upload/c_fill,f_auto,fl_progressive,g_center,h_675,pg_1,q_80,w_1200/cd4c128b4182d7b2fba8152d7bb35733.jpg',
   'publishedAt': '2022-05-23T21:35:00Z',
   'content': 'GameStop has officially thrown itself headlong into the web3 vipers nest with a new app release, though its hard to say whether its proposed population of gamers and game developers will take up the … [+3255 chars]'}

In [61]:
# Function for getting sentiment score DF's 
def get_sentiment_score(article, based_on):
    sentiments = []
    for article in article["articles"]:
        try:
            text = article[based_on]
            date = article["publishedAt"][:10]
            sentiment = analyzer.polarity_scores(text)
            compound = sentiment["compound"]
            pos = sentiment["pos"]
            neu = sentiment["neu"]
            neg = sentiment["neg"]
            sentiments.append({
                based_on : text,
                "Compound": compound,
                "Negative": neg,
                "Neutral": neu,
                "Positive": pos
            })
        except AttributeError:
            pass
    df = pd.DataFrame(sentiments)
    return df

In [62]:
# Create the Bitcoin sentiment scores DataFrame
btc_sent_df = get_sentiment_score(btc_articles, "content")
btc_sent_df

content  Compound  Negative  \
0   New York lawmakers have passed a bill\r\n that...   -0.5574     0.098   
1   Rapper and entrepreneur Shawn Carter, better k...    0.4404     0.000   
2   A new study on bitcoin calls into question whe...    0.5267     0.000   
3   Image caption, President Faustin-Archange Toua...    0.5106     0.000   
4   You can now reportedly pay for your burritos a...   -0.1027     0.040   
5   By Joe TidyCyber reporter \r\nCryptocurrencies...    0.2960     0.074   
6   (CNN)El Salvador has embraced Bitcoin like no ...    0.1027     0.046   
7   As a kid, I remember when my father tried to u...    0.3818     0.052   
8   Customers at Chipotle will now be able to pay ...    0.3182     0.040   
9   Celsius has not said what it plans to do next\...    0.0000     0.000   
10  Photo Illustration by Grayson Blackmon / The V...   -0.4767     0.103   
11  June 6 (Reuters) - Bitcoin rose 5.2% to $31,44...    0.0000     0.000   
12  June 13 (Reuters) - Bitcoin dropped 12.1% to $...   -0.3818     0.075   
13  SINGAPORE, June 14 (Reuters) - Bitcoin neared ...    0.0000     0.000   
14  May 24 (Reuters) - The crypto winter is into i...    0.1326     0.000   
15  After two linked cryptocurrencies collapsed an...   -0.8593     0.300   
16  Terra launched a new blockchain with a new lun...   -0.2960     0.084   
17  Terra is launching a new blockchain with a new...   -0.4019     0.083   
18  Sam Bankman-Fried, founder of the digital asse...    0.4051     0.059   
19  As cryptocurrencies have grown in value, Bitco...    0.2960     0.099   

    Neutral  Positive  
0     0.902     0.000  
1     0.923     0.077  
2     0.876     0.124  
3     0.836     0.164  
4     0.960     0.000  
5     0.792     0.134  
6     0.867     0.087  
7     0.833     0.114  
8     0.883     0.077  
9     1.000     0.000  
10    0.897     0.000  
11    1.000     0.000  
12    0.925     0.000  
13    1.000     0.000  
14    0.956     0.044  
15    0.646     0.054  
16    0.838     0.079  
17    0.879     0.039  
18    0.809     0.132  
19    0.714     0.187

In [63]:
# Create the Ethereum sentiment scores DataFrame
eth_sent_df = get_sentiment_score(eth_articles, "content")
eth_sent_df

content  Compound  Negative  \
0   GameStop has officially thrown itself headlong...   -0.1027     0.040   
1   GameStop is going all-in on crypto. The video ...    0.1280     0.000   
2   The ability to conduct external transfers on P...    0.3182     0.000   
3   A decentralized autonomous organization, or DA...    0.5859     0.000   
4   Crypto Winter It May Be, But Ethereum Looks Li...    0.3612     0.044   
5   New York lawmakers have passed a bill\r\n that...   -0.5574     0.098   
6   DAVOS, Switzerland, May 25 (Reuters) - Ethereu...    0.0258     0.000   
7   When Nvidia launched its Ampere Lite Hash Rate...   -0.3818     0.085   
8   The crypto roller coaster ride continues to pl...    0.0000     0.000   
9   June 6 (Reuters) - Bitcoin rose 5.2% to $31,44...    0.0000     0.000   
10  June 13 (Reuters) - Bitcoin dropped 12.1% to $...   -0.3818     0.075   
11  [In a tweet] Buterin shared a paper by AI theo...   -0.4404     0.119   
12  YouTube is being criticised for failing to tac...   -0.9485     0.372   
13  The cryptocurrency boom over the past few year...   -0.1531     0.045   
14  We first heard about the potential of blockcha...    0.4404     0.000   
15  Every time you turn around, it seems another c...    0.4215     0.000   
16  NFT marketplaces are becoming the trend, and e...    0.1280     0.000   
17  Solana, the third largest cryptocurrency by ma...   -0.4767     0.100   
18  The cryptocurrency boom over the past few year...   -0.1531     0.045   
19  6 with 6 posters participating\r\nNew York's s...    0.4404     0.000   

    Neutral  Positive  
0     0.960     0.000  
1     0.954     0.046  
2     0.941     0.059  
3     0.866     0.134  
4     0.875     0.081  
5     0.902     0.000  
6     0.966     0.034  
7     0.847     0.069  
8     1.000     0.000  
9     1.000     0.000  
10    0.925     0.000  
11    0.786     0.095  
12    0.628     0.000  
13    0.955     0.000  
14    0.925     0.075  
15    0.917     0.083  
16    0.958     0.042  
17    0.900     0.000  
18    0.955     0.000  
19    0.878     0.122

In [64]:
# Describe the Bitcoin Sentiment
btc_sent_df.describe()

Compound   Negative    Neutral   Positive
count  20.000000  20.000000  20.000000  20.000000
mean    0.016715   0.057650   0.876800   0.065600
std     0.398865   0.068859   0.092058   0.061384
min    -0.859300   0.000000   0.646000   0.000000
25%    -0.317450   0.000000   0.835250   0.000000
50%     0.051350   0.049000   0.881000   0.065500
75%     0.334100   0.083250   0.932750   0.116500
max     0.526700   0.300000   1.000000   0.187000

In [65]:
# Describe the Ethereum Sentiment
eth_sent_df.describe()

Compound   Negative    Neutral  Positive
count  20.000000  20.000000  20.000000  20.00000
mean   -0.037305   0.051150   0.906900   0.04200
std     0.405662   0.085919   0.084171   0.04527
min    -0.948500   0.000000   0.628000   0.00000
25%    -0.381800   0.000000   0.877250   0.00000
50%     0.000000   0.020000   0.925000   0.03800
75%     0.328950   0.077500   0.955750   0.07650
max     0.585900   0.372000   1.000000   0.13400

### Questions:

Q: Which coin had the highest mean positive score?

A: Bitcoin 0.065600

Q: Which coin had the highest compound score?

A: Ethereum 0.585900

Q. Which coin had the highest positive score?

A: Bitcoin 0.187000

---

## 2. Natural Language Processing
---
###   Tokenizer

In this section, you will use NLTK and Python to tokenize the text for each coin. Be sure to:
1. Lowercase each word.
2. Remove Punctuation.
3. Remove Stopwords.

In [66]:
from nltk.tokenize import word_tokenize, sent_tokenize
from nltk.corpus import stopwords
from nltk.stem import WordNetLemmatizer, PorterStemmer
from string import punctuation
import re

In [77]:
# Instantiate the lemmatizer
lemmatizer = WordNetLemmatizer()

# Create a list of stopwords
sw = set(stopwords.words("english"))

# Expand the default stopwords list if necessary
# YOUR CODE HERE!

In [78]:
# Complete the tokenizer function
def tokenizer(text):
    """Tokenizes text."""
    
    # Remove the punctuation from text
    regex = re.compile("[^a-zA-Z0-9 ]")
    re_clean = regex.sub('', text)
   
    # Create a tokenized list of the words
    words = word_tokenize(re_clean.lower())
    
    #lemmatizer = WordNetLemmatizer()
    
    # Lemmatize words into root words
    lem = [lemmatizer.lemmatize(word) for word in words]
    
   
    # Convert the words to lowercase
    words = [word for word in words if word not in sw]
    
    # Remove the stop words
    tokens = [word.lower() for word in lem if word.lower() not in sw.union(sw)]

    return tokens



In [69]:
btc_sent_df.iloc[0]['content']

"New York lawmakers have passed a bill\r\n that would temporarily ban new bitcoin\r\n mining operations. Early on Friday, state senators voted 36-27 to pass the legislation. It's now bound for the desk of… [+2036 chars]"

In [79]:
# Create a new tokens column for Bitcoin
btc_sent_df['tokens'] = btc_sent_df['content'].apply(tokenizer)
btc_sent_df

content  Compound  Negative  \
0   New York lawmakers have passed a bill\r\n that...   -0.5574     0.098   
1   Rapper and entrepreneur Shawn Carter, better k...    0.4404     0.000   
2   A new study on bitcoin calls into question whe...    0.5267     0.000   
3   Image caption, President Faustin-Archange Toua...    0.5106     0.000   
4   You can now reportedly pay for your burritos a...   -0.1027     0.040   
5   By Joe TidyCyber reporter \r\nCryptocurrencies...    0.2960     0.074   
6   (CNN)El Salvador has embraced Bitcoin like no ...    0.1027     0.046   
7   As a kid, I remember when my father tried to u...    0.3818     0.052   
8   Customers at Chipotle will now be able to pay ...    0.3182     0.040   
9   Celsius has not said what it plans to do next\...    0.0000     0.000   
10  Photo Illustration by Grayson Blackmon / The V...   -0.4767     0.103   
11  June 6 (Reuters) - Bitcoin rose 5.2% to $31,44...    0.0000     0.000   
12  June 13 (Reuters) - Bitcoin dropped 12.1% to $...   -0.3818     0.075   
13  SINGAPORE, June 14 (Reuters) - Bitcoin neared ...    0.0000     0.000   
14  May 24 (Reuters) - The crypto winter is into i...    0.1326     0.000   
15  After two linked cryptocurrencies collapsed an...   -0.8593     0.300   
16  Terra launched a new blockchain with a new lun...   -0.2960     0.084   
17  Terra is launching a new blockchain with a new...   -0.4019     0.083   
18  Sam Bankman-Fried, founder of the digital asse...    0.4051     0.059   
19  As cryptocurrencies have grown in value, Bitco...    0.2960     0.099   

    Neutral  Positive                                             tokens  
0     0.902     0.000  [new, york, lawmaker, passed, bill, would, tem...  
1     0.923     0.077  [rapper, entrepreneur, shawn, carter, better, ...  
2     0.876     0.124  [new, study, bitcoin, call, question, whether,...  
3     0.836     0.164  [image, caption, president, faustinarchange, t...  
4     0.960     0.000  [reportedly, pay, burrito, taco, bitcoin, digi...  
5     0.792     0.134  [joe, tidycyber, reporter, cryptocurrencies, c...  
6     0.867     0.087  [cnnel, salvador, ha, embraced, bitcoin, like,...  
7     0.833     0.114  [kid, remember, father, tried, use, broom, han...  
8     0.883     0.077  [customer, chipotle, able, pay, burrito, crypt...  
9     1.000     0.000  [celsius, ha, said, plan, nextexhibition, expe...  
10    0.897     0.000  [photo, illustration, grayson, blackmon, verge...  
11    1.000     0.000  [june, 6, reuters, bitcoin, rose, 52, 3144176,...  
12    0.925     0.000  [june, 13, reuters, bitcoin, dropped, 121, 233...  
13    1.000     0.000  [singapore, june, 14, reuters, bitcoin, neared...  
14    0.956     0.044  [may, 24, reuters, crypto, winter, ninth, week...  
15    0.646     0.054  [two, linked, cryptocurrencies, collapsed, bec...  
16    0.838     0.079  [terra, launched, new, blockchain, new, luna, ...  
17    0.879     0.039  [terra, launching, new, blockchain, new, luna,...  
18    0.809     0.132  [sam, bankmanfried, founder, digital, asset, e...  
19    0.714     0.187  [cryptocurrencies, grown, value, bitcoin, mini...

In [80]:
# Create a new tokens column for Ethereum
eth_sent_df['tokens'] = eth_sent_df['content'].apply(tokenizer)
eth_sent_df

content  Compound  Negative  \
0   GameStop has officially thrown itself headlong...   -0.1027     0.040   
1   GameStop is going all-in on crypto. The video ...    0.1280     0.000   
2   The ability to conduct external transfers on P...    0.3182     0.000   
3   A decentralized autonomous organization, or DA...    0.5859     0.000   
4   Crypto Winter It May Be, But Ethereum Looks Li...    0.3612     0.044   
5   New York lawmakers have passed a bill\r\n that...   -0.5574     0.098   
6   DAVOS, Switzerland, May 25 (Reuters) - Ethereu...    0.0258     0.000   
7   When Nvidia launched its Ampere Lite Hash Rate...   -0.3818     0.085   
8   The crypto roller coaster ride continues to pl...    0.0000     0.000   
9   June 6 (Reuters) - Bitcoin rose 5.2% to $31,44...    0.0000     0.000   
10  June 13 (Reuters) - Bitcoin dropped 12.1% to $...   -0.3818     0.075   
11  [In a tweet] Buterin shared a paper by AI theo...   -0.4404     0.119   
12  YouTube is being criticised for failing to tac...   -0.9485     0.372   
13  The cryptocurrency boom over the past few year...   -0.1531     0.045   
14  We first heard about the potential of blockcha...    0.4404     0.000   
15  Every time you turn around, it seems another c...    0.4215     0.000   
16  NFT marketplaces are becoming the trend, and e...    0.1280     0.000   
17  Solana, the third largest cryptocurrency by ma...   -0.4767     0.100   
18  The cryptocurrency boom over the past few year...   -0.1531     0.045   
19  6 with 6 posters participating\r\nNew York's s...    0.4404     0.000   

    Neutral  Positive                                             tokens  
0     0.960     0.000  [gamestop, ha, officially, thrown, headlong, w...  
1     0.954     0.046  [gamestop, going, allin, crypto, video, game, ...  
2     0.941     0.059  [ability, conduct, external, transfer, paypals...  
3     0.866     0.134  [decentralized, autonomous, organization, dao,...  
4     0.875     0.081  [crypto, winter, may, ethereum, look, like, bu...  
5     0.902     0.000  [new, york, lawmaker, passed, bill, would, tem...  
6     0.966     0.034  [davos, switzerland, may, 25, reuters, ethereu...  
7     0.847     0.069  [nvidia, launched, ampere, lite, hash, rate, l...  
8     1.000     0.000  [crypto, roller, coaster, ride, continues, plu...  
9     1.000     0.000  [june, 6, reuters, bitcoin, rose, 52, 3144176,...  
10    0.925     0.000  [june, 13, reuters, bitcoin, dropped, 121, 233...  
11    0.786     0.095  [tweet, buterin, shared, paper, ai, theorist, ...  
12    0.628     0.000  [youtube, criticised, failing, tackle, network...  
13    0.955     0.000  [cryptocurrency, boom, past, year, ha, helped,...  
14    0.925     0.075  [first, heard, potential, blockchain, supply, ...  
15    0.917     0.083  [every, time, turn, around, seems, another, ce...  
16    0.958     0.042  [nft, marketplace, becoming, trend, every, one...  
17    0.900     0.000  [solana, third, largest, cryptocurrency, marke...  
18    0.955     0.000  [cryptocurrency, boom, past, year, ha, helped,...  
19    0.878     0.122  [6, 6, poster, participatingnew, york, state, ...

---

### NGrams and Frequency Analysis

In this section you will look at the ngrams and word frequency for each coin. 

1. Use NLTK to produce the n-grams for N = 2. 
2. List the top 10 words for each coin. 

In [81]:
from collections import Counter
from nltk import ngrams

In [82]:
# Get token function
def get_token(df):
    tokens = []
    for i in df['tokens']:
        tokens.extend(i)
    return tokens

btc_token = get_token(btc_sent_df)
eth_token = get_token(eth_sent_df)

In [83]:
# N-gram Function
def bigram_counter(tokens, N):
    words_count = dict(Counter(ngrams(tokens, n=N)))
    return words_count

In [84]:
# Generate the Bitcoin N-grams where N=2
btc_bigram = bigram_counter(btc_token, 2)
btc_bigram

{('new', 'york'): 1,
 ('york', 'lawmaker'): 1,
 ('lawmaker', 'passed'): 1,
 ('passed', 'bill'): 1,
 ('bill', 'would'): 1,
 ('would', 'temporarily'): 1,
 ('temporarily', 'ban'): 1,
 ('ban', 'new'): 1,
 ('new', 'bitcoin'): 1,
 ('bitcoin', 'mining'): 2,
 ('mining', 'operation'): 1,
 ('operation', 'early'): 1,
 ('early', 'friday'): 1,
 ('friday', 'state'): 1,
 ('state', 'senator'): 1,
 ('senator', 'voted'): 1,
 ('voted', '3627'): 1,
 ('3627', 'pas'): 1,
 ('pas', 'legislation'): 1,
 ('legislation', 'bound'): 1,
 ('bound', 'desk'): 1,
 ('desk', '2036'): 1,
 ('2036', 'char'): 1,
 ('char', 'rapper'): 1,
 ('rapper', 'entrepreneur'): 1,
 ('entrepreneur', 'shawn'): 1,
 ('shawn', 'carter'): 1,
 ('carter', 'better'): 1,
 ('better', 'known'): 1,
 ('known', 'jayz'): 1,
 ('jayz', 'bringing'): 1,
 ('bringing', 'bitcoin'): 1,
 ('bitcoin', 'place'): 1,
 ('place', 'grew'): 1,
 ('grew', 'thursday'): 1,
 ('thursday', 'jayz'): 1,
 ('jayz', 'former'): 1,
 ('former', 'twitter'): 1,
 ('twitter', 'ceo'): 1,
 ('c

In [85]:
# Generate the Ethereum N-grams where N=2
eth_bigram = bigram_counter(eth_token, 2)
eth_bigram

{('gamestop', 'ha'): 1,
 ('ha', 'officially'): 1,
 ('officially', 'thrown'): 1,
 ('thrown', 'headlong'): 1,
 ('headlong', 'web3'): 1,
 ('web3', 'viper'): 1,
 ('viper', 'nest'): 1,
 ('nest', 'new'): 1,
 ('new', 'app'): 1,
 ('app', 'release'): 1,
 ('release', 'though'): 1,
 ('though', 'hard'): 1,
 ('hard', 'say'): 1,
 ('say', 'whether'): 1,
 ('whether', 'proposed'): 1,
 ('proposed', 'population'): 1,
 ('population', 'gamers'): 1,
 ('gamers', 'game'): 1,
 ('game', 'developer'): 1,
 ('developer', 'take'): 1,
 ('take', '3255'): 1,
 ('3255', 'char'): 1,
 ('char', 'gamestop'): 1,
 ('gamestop', 'going'): 1,
 ('going', 'allin'): 1,
 ('allin', 'crypto'): 1,
 ('crypto', 'video'): 1,
 ('video', 'game'): 1,
 ('game', 'retailer'): 1,
 ('retailer', 'launchedits'): 1,
 ('launchedits', 'selfcustodial'): 1,
 ('selfcustodial', 'ethereum'): 1,
 ('ethereum', 'digital'): 1,
 ('digital', 'wallet'): 1,
 ('wallet', 'said'): 1,
 ('said', 'monday'): 1,
 ('monday', 'wallet'): 1,
 ('wallet', 'accessible'): 1,
 ('a

In [86]:
# Function token_count generates the top 10 words for a given coin
def token_count(tokens, N=3):
    """Returns the top N tokens from the frequency count"""
    return Counter(tokens).most_common(N)

In [87]:
# Use token_count to get the top 10 words for Bitcoin
token_count(btc_token, 10)

[('char', 20),
 ('bitcoin', 15),
 ('new', 8),
 ('token', 6),
 ('blockchain', 6),
 ('digital', 4),
 ('biggest', 4),
 ('cryptocurrencies', 4),
 ('cryptocurrency', 4),
 ('reuters', 4)]

In [88]:
# Use token_count to get the top 10 words for Ethereum
token_count(eth_token, 10)

[('char', 20),
 ('cryptocurrency', 10),
 ('ha', 6),
 ('crypto', 5),
 ('new', 4),
 ('ethereum', 4),
 ('market', 4),
 ('year', 4),
 ('nft', 4),
 ('video', 3)]

---

### Word Clouds

In this section, you will generate word clouds for each coin to summarize the news for each coin

In [89]:
from wordcloud import WordCloud
import matplotlib.pyplot as plt
plt.style.use('seaborn-whitegrid')
import matplotlib as mpl
mpl.rcParams['figure.figsize'] = [20.0, 10.0]

ModuleNotFoundError: No module named 'wordcloud'

In [ ]:
# Generate the Bitcoin word cloud
# YOUR CODE HERE!

In [ ]:
# Generate the Ethereum word cloud
# YOUR CODE HERE!

---
## 3. Named Entity Recognition

In this section, you will build a named entity recognition model for both Bitcoin and Ethereum, then visualize the tags using SpaCy.

In [ ]:
import spacy
from spacy import displacy

In [ ]:
# Download the language model for SpaCy
# !python -m spacy download en_core_web_sm

In [ ]:
# Load the spaCy model
nlp = spacy.load('en_core_web_sm')

---
### Bitcoin NER

In [ ]:
# Concatenate all of the Bitcoin text together
# YOUR CODE HERE!

In [ ]:
# Run the NER processor on all of the text
# YOUR CODE HERE!

# Add a title to the document
# YOUR CODE HERE!

In [ ]:
# Render the visualization
# YOUR CODE HERE!

In [ ]:
# List all Entities
# YOUR CODE HERE!

---

### Ethereum NER

In [ ]:
# Concatenate all of the Ethereum text together
# YOUR CODE HERE!

In [ ]:
# Run the NER processor on all of the text
# YOUR CODE HERE!

# Add a title to the document
# YOUR CODE HERE!

In [ ]:
# Render the visualization
# YOUR CODE HERE!

In [ ]:
# List all Entities
# YOUR CODE HERE!

---